airflow/contrib/sensors/spark_kubernetes_sensor.py

```diff
@@ -0,0 +1,82 @@
+# -*- coding: utf-8 -*-
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+from typing import Dict, Optional
+
+from kubernetes import client
+
+from airflow.contrib.hooks.kubernetes_hook import Kuberneteshook
+from airflow.exceptions import AirflowException
+from airflow.sensors.base_sensor_operator import BaseSensorOperator
+from airflow.utils.decorators import apply_defaults
+
+
+class SparkKubernetesSensor(BaseSensorOperator):
+    """
+    checks sparkapplication state on kubernetes
+    :param sparkapplication_name: sparkapplication resource name
+    :param namespace: the kubernetes namespace where the sparkApplication reside in
+    :param kube_config: kubeconfig file location
+    :param in_cluster: if airflow runs inside kubernetes pod take configuration from inside the cluster
+    """
+
+    template_fields = ('sparkapplication_name', 'namespace', 'kube_config')
+    INTERMEDIATE_STATES = ('SUBMITTED', 'RUNNING',)
+    FAILURE_STATES = ('FAILED', 'SUBMISSION_FAILED', 'UNKNOWN')
+    SUCCESS_STATES = 'COMPLETED'
+
+    @apply_defaults
+    def __init__(self,
+                 sparkapplication_name: str,
+                 namespace: str = 'default',
+                 kube_config: Optional[str] = None,
+                 in_cluster: bool = False,
+                 *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.sparkapplication_name = sparkapplication_name
+        self.namespace = namespace
+        self.kube_config = kube_config
+        self.in_cluster = in_cluster
+        self.api_client = Kuberneteshook(
```

Comment:
Hook should not be initialized in the constructor but in the execute method. The constructor should be as light as possible, because this causes unnecessary load. DAG is loaded many times in various components, and the hook is needed only when executing the execute method.

Reply:
I initialized the hook in the constructor so the poke method that can run many times in the sensor won't need to get a new connection each time it poke's. it's ok?

Reply:
Maybe it's worth moving it to the pre_execute method?  The constructor is called by every worker repeatedly when each task is started. It really should be low cost.

Reply:
there is no pre_execute method for a sensor, I moved it to the poke method